GameEngine/Agents/GreedyBestFirstAgent.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace GameEngine.Agents
+{
+    public class GreedyBestFirstAgent : IAgent
+    {
+        public Queue<SearchNode> solution;
+
+        public Play FormulatePlay(GameState state)
+        {
+            solution = solution ?? GreedyBestFirstSearch(state);
+            var nextNode = solution.Dequeue();
+            if (nextNode.Parent.State != state)
+            {
+                throw new StateMismatchException();
+            }
+            return nextNode.Action;
+        }
+
+        private Queue<SearchNode> GreedyBestFirstSearch(GameState state)
+        {
+            var frontierNodes = new List<SearchNode> { new RootNode(state) };
+            while (true)
+            {
+                if (frontierNodes.Count == 0)
+                {
+                    throw new NoSolutionFoundException();
+                }
+                var node = frontierNodes.RemoveMin(H);
+                if ( H(node) == 0 )
+                {
+                    return new Queue<SearchNode>(node.Solution());
+                }
+                frontierNodes.AddRange(node.Expand());
+            }
+        }
+
+        private int H(SearchNode node)
```

Comment:
❌  Heuristics are functions of the game state rather than the search algorithm.   You're going to want to be able to evaluate multiple algorithms while controlling for the informedness of the heuristics they use to solve the problem.

The ideal case in my mind is that the search could be parameterized on the heuristic to use.  If that's not possible or is too against the language idiom, I don't know what the right solution is.  One option is generating N different child classes of an abstract parent, each with a specific grounding of the heuristic.

Reply:
I think, because they're the same in this case, I'm conflating _f(n)_ and _h(n)_.

Architecturally speaking, this is the part of this changeset that I feel least sure about. There were a lot of options for how to map the highly functional style from the text into an object oriented style.

For instance, I thought about abstracting out the simple tree search algorithm from `GreedyBestFirstSearch(GameState state)`. Calling `RemoveMin` seems like it might be called the "strategy". `SearchNode Strategy(List<SearchNode> frontierNodes)` could be a required function of an abstract class `TreeSearchAgent` that `GreedyBestFirstAgent` implements. Maybe the constructor for heuristic search agents can require a parameterized heuristic function of `GameState`. We have options.

Also, to some extent, I'd like to let the development of additional agents drive the particulars of what gets commonized and what doesn't. Let's chat.

Reply:
I've abstracted `TreeSearchAgent` out to handle caching the solution and executing a standard tree search. It has a single required abstract function, `RemoveStrategicNode`. `GreedyBestFirstAgent` implements this to choose the first node with the lowest value of _h(n)_. The specific heuristic comes from the generic type used when initializing the agent.

I think this should satisfy our ongoing needs. I'm going to implement a few more agents that use tree search and see where we are.